### ISSUE TITLE
Averaging Weights Not Working 

### ISSUE THREAD

**Issue by HunterMorera** [ultralytics/yolov5#11667]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am trying to average weights from multiple .pt files to perform an ensemble of sorts. I have coded up a script to do this, however, the avg weights do not produce any predictions on the test data. I have checked the weights after saving and they appear to be correct. Any Suggestions?

Please see script below:

```python

import torch
import os
import copy


m1 = torch.load(os.path.join("weight_files", "last.pt"), map_location=torch.device('cpu'))
m2 = torch.load(os.path.join("weight_files","epoch480.pt"), map_location=torch.device('cpu'))
m3 = torch.load(os.path.join("weight_files","epoch460.pt"), map_location=torch.device('cpu'))
mF = torch.load(os.path.join("weight_files","epoch0.pt"), map_location=torch.device('cpu'))



m1w, m2w, m3w, mFw = m1['model'].state_dict(), m2['model'].state_dict(), m3['model'].state_dict(), mF['model'].state_dict()

for key in mFw:

		new_weight = torch.sum(torch.stack([m1w[key].double(), m2w[key], m3w[key]]), dim=0)

		mFw[key].copy_(torch.div(new_weight, 3.0))


mF['model'].state_dict().update(mFw)


torch.save(mF ,os.path.join("weight_files", "avg.pt"))

```

**Comment by glenn-jocher** [comment]
@HunterMorera hello,

Thank you for reaching out and providing the code snippet. Your code looks good and it should work as expected. 

If your weights are correctly saved in the "avg.pt" file, the inference should work just like with a single "last.pt" file. The script provided looks reasonable and I suggest making sure that your weights path is correct and that your data is in the right format for the inference.

Please let me know if there is anything else I can help you with.

Best, 
Glenn Jocher

**Comment by HunterMorera** [comment]
Is there anything else that would need to be changed in the .pt file besides these weights? I am using epoch0 as a placeholder dictionary to then copy the averaged values into. Not sure if there is any other values that should be averaged?

**Comment by lucastack** [comment]
Hello @HunterMorera, are you using `detect.py` to run detections? Have you checked the output of the model?

**Comment by HunterMorera** [comment]
@lucastack I am using detect.py to run detections. It just says no predictions for every image in the test set. One issue was overflow when adding all of the weights so I had to cast to double.

I tried loading the same weight file last.py for all 3 and dividing by 3 to get the same model back. The weights are the same as I have checked them with torch.eq() however detect.py says no detections on the test images.

**Comment by glenn-jocher** [comment]
@HunterMorera it's good to hear that you have checked the weights and they are the same. I suggest you try checking the output of the model using `--save-txt` flag in `detect.py` and verify that the output format is correct. Also, make sure that there is no issue with your test data.

If this doesn't help, you can try some other ways to perform ensemble, such as weighted averaging of the predictions from different models. 

Please let me know if this helps or if there is anything else I can assist you with.

Best,
Glenn Jocher

**Comment by lucastack** [comment]
@HunterMorera Ok! if you are using that script, can you try to print out the result of the model by adding `print(pred)` just after `pred = model(im, augment=augment, visualize=visualize)` ? Here:

![image](https://github.com/ultralytics/yolov5/assets/53802157/766534a4-d567-4311-ba24-1e3a52793a18)


Also, are your individual models capable of recognize the objects?

Regards!

**Comment by glenn-jocher** [comment]
@lucastack

As a follow-up to the issue you reported, I suggest printing out the result of the model by adding `print(pred)` just after `pred = model(im, augment=augment, visualize=visualize)` in `detect.py`. This will help you verify if the model is producing any results.

Also, it would be helpful to confirm whether your individual models are capable of recognizing the objects in the images used for testing.

Please let me know if there is anything else I can assist you with.

Regards!

**Comment by HunterMorera** [comment]
Interesting finding. I tried loading in the same model 3 times to debug and I found the following results.

```python
import torch
import os
import copy


m1 = torch.load(os.path.join("weight_files", "last.pt"), map_location=torch.device('cpu'))
m2 = torch.load(os.path.join("weight_files","last.pt.pt"), map_location=torch.device('cpu'))
m3 = torch.load(os.path.join("weight_files","last.pt.pt"), map_location=torch.device('cpu'))
mF = torch.load(os.path.join("weight_files","epoch0.pt"), map_location=torch.device('cpu'))



m1w, m2w, m3w, mFw = m1['model'].state_dict(), m2['model'].state_dict(), m3['model'].state_dict(), mF['model'].state_dict()

for key in mFw:

		new_weight = torch.sum(torch.stack([m1w[key].double(), m2w[key], m3w[key]]), dim=0)

		mFw[key].copy_(torch.div(new_weight, 3.0))


mF['model'].state_dict().update(mFw)


torch.save(mF ,os.path.join("weight_files", "avg.pt"))
```
The above script when inferring with avg.pt provides no detections however, if we swap out the epoch0.pt to last.pt we get the correct output. However, if we exchange it for another epoch like epoch480.pt it outputs the results as if epoch480.pt was being used. Thus no matter what it is using the weights from the mF model that's loaded in at the begging of the script. Not sure if there is anything I am missing in that mF is not being updated despite the state dict being updated and the weights appearing the same when you read them.

Does the detect.py use other things like optimizer state or something?

**Comment by lucastack** [comment]
@HunterMorera I dont think the optimizer state has to do anything with this. As you pointed out, the problem is that mF weights are not being updated. And in that case, of course the model is not going to be able to detect anything because it's just the weights from epoch 0. 

Have you tried just replacing the values instead of copying them? I mean `mFw[key] = torch.div(new_weight, 3.0)`

Regards!

**Comment by HunterMorera** [comment]
@lucastack @glenn-jocher 

Did some more debugging and figured it out. I believe its a quark with the YOLOv5 detect.py code. I debugged what other part of the model was needed to be correct besides the weights. It was the ema value. It needs to be set to None for the model to infer correctly. Its now working as expected. So correct code is:

```python
import torch
import os
import copy


m1 = torch.load(os.path.join("weight_files", "last.pt"), map_location=torch.device('cpu'))
m2 = torch.load(os.path.join("weight_files","last.pt.pt"), map_location=torch.device('cpu'))
m3 = torch.load(os.path.join("weight_files","last.pt.pt"), map_location=torch.device('cpu'))
mF = torch.load(os.path.join("weight_files","epoch0.pt"), map_location=torch.device('cpu'))



m1w, m2w, m3w, mFw = m1['model'].state_dict(), m2['model'].state_dict(), m3['model'].state_dict(), mF['model'].state_dict()

for key in mFw:

		new_weight = torch.sum(torch.stack([m1w[key].double(), m2w[key], m3w[key]]), dim=0)

		mFw[key].copy_(torch.div(new_weight, 3.0))


mF['model'].state_dict().update(mFw)
mF['ema'] = None


torch.save(mF ,os.path.join("weight_files", "avg.pt"))

**Comment by glenn-jocher** [comment]
Hello @lucastack,

Thank you for the update on your issue. It's good to hear that you figured out the problem with your code. Indeed, the `ema` value needs to be set to `None` for the model to infer correctly as stated in the implementation code.

If you have any further questions or need any more assistance, please do not hesitate to reach out. We are always happy to help.

Best regards.
